src/ray/common/ray_syncer/ray_syncer.cc

```diff
@@ -124,234 +146,210 @@ std::array<int64_t, kComponentArraySize> &NodeSyncConnection::GetNodeComponentVe
-    stub_->async()->Update(
-        client_context.get(),
-        request,
-        response,
-        [arena, client_context, holder = std::move(holder)](grpc::Status status) {
-          if (!status.ok()) {
-            RAY_LOG(ERROR) << "Sending request failed because of "
-                           << status.error_message();
-          }
-        });
-  }
-}
-
-ServerSyncConnection::ServerSyncConnection(
+RayServerBidiReactor::RayServerBidiReactor(
+    grpc::CallbackServerContext *server_context,
     instrumented_io_context &io_context,
-    const std::string &remote_node_id,
-    std::function<void(std::shared_ptr<RaySyncMessage>)> message_processor)
-    : NodeSyncConnection(io_context, remote_node_id, std::move(message_processor)) {}
-
-ServerSyncConnection::~ServerSyncConnection() {
-  // If there is a pending request, we need to cancel it. Otherwise, rpc will
-  // hang there forever.
-  while (!unary_reactors_.empty()) {
-    unary_reactors_.back()->Finish(grpc::Status::CANCELLED);
-    unary_reactors_.pop_back();
-  }
+    const std::string &local_node_id,
+    std::function<void(std::shared_ptr<const RaySyncMessage>)> message_processor,
+    std::function<void(const std::string &, bool)> cleanup_cb)
+    : BidiReactor<ServerBidiReactor>(io_context,
+                                     GetNodeIDFromServerContext(server_context),
+                                     std::move(message_processor),
+                                     std::move(cleanup_cb)),
+      server_context_(server_context) {
+  // Send the local node id to the remote
+  server_context_->AddInitialMetadata("node_id", NodeID::FromBinary(local_node_id).Hex());
+  StartSendInitialMetadata();
+
+  // Start pulling from remote
+  StartPull();
 }
 
-void ServerSyncConnection::HandleLongPollingRequest(grpc::ServerUnaryReactor *reactor,
-                                                    RaySyncMessages *response) {
-  unary_reactors_.push_back(reactor);
-  responses_.push_back(response);
-}
+void RayServerBidiReactor::Disconnect() { Finish(grpc::Status::OK); }
 
-void ServerSyncConnection::DoSend() {
-  // There is no receive request
-  if (unary_reactors_.empty() || sending_buffer_.empty()) {
-    return;
-  }
+void RayServerBidiReactor::OnCancel() { Disconnect(); }
 
-  RAY_CHECK(!responses_.empty());
+void RayServerBidiReactor::OnDone() {
+  io_context_.dispatch(
+      [this]() {
+        cleanup_cb_(GetRemoteNodeID(), false);
+        delete this;
+      },
+      "");
+}
 
-  size_t message_bytes = 0;
-  auto iter = sending_buffer_.begin();
-  while (message_bytes < RayConfig::instance().max_sync_message_batch_bytes() &&
-         iter != sending_buffer_.end()) {
-    message_bytes += iter->second->sync_message().size();
-    // TODO (iycheng): Use arena allocator for optimization
-    responses_.back()->add_sync_messages()->CopyFrom(*iter->second);
-    sending_buffer_.erase(iter++);
-  }
+RayClientBidiReactor::RayClientBidiReactor(
```

Comment:
let's add unit tests for Server/Client Reactor? this will also help understand how the Reactor is used.

Reply:
let me give it a try